File: app/src/main/java/com/softaai/designpatterns/java/structural/proxy/ProxyImage.java
package com.softaai.designpatterns.java.structural.proxy;

/**
 * Created by Amol Pawar on 03-04-2019.
 * softAai Apps
 */
public class ProxyImage implements Image {
    private RealImage realImage;
    private String fileName;
    public ProxyImage(String fileName){
        this.fileName = fileName;
    }

    @Override
    public void display() {
        if(realImage == null){
            realImage = new RealImage(fileName);
        }
        realImage.display();
    }
}

File: app/src/main/java/com/softaai/designpatterns/kotlin/structural/Proxy/ProxyImage.kt
package com.softaai.designpatterns.kotlin.structural.Proxy


/**
 * Created by Amol Pawar on 03-04-2019.
 * softAai Apps
 */
class ProxyImage(fileName:String) : Image {

    private var realImage:RealImage? = null
    private var fileName :String

    init {
        this.fileName = fileName
    }
    override fun display() {
        if (realImage == null){
            realImage = RealImage(fileName)
        }

        realImage?.display()
    }
}